### ISSUE TITLE
Service fails to restart on device restart

### ISSUE THREAD

**Issue by safi-manar** [qtmdev/AndroidInference#21]
**Build:** 1b94548dc85df5a45f53ed188631d13e07de3dbd
**UUID:** `4803196c-3dcf-4a7d-9f7d-e74ab9a6e010`

**Deescription:** The service does not restart upon a system restart and so data collection no longer persists.

**To Replicate:** 
1. Install a fresh copy of 1b94548dc85df5a45f53ed188631d13e07de3dbd and connect device to AndroidMonitor.
2. Proceed with the procedure of completing the survey.
3. Once alarms have been set, close the app (double tap and swipe the app to close)
4. Do not unplug the device. Turn off the phone. Turn on the phone.
5. The error should occur when the device is unlocked. The service is unable to restart and the device no longer collects and sends data.

**Error:** 

``````
06-08 15:04:44.811 2767-2767/us.michaelchen.compasslogger W/System: ClassLoader referenced unknown path: /data/app/us.michaelchen.compasslogger-1/lib/arm
06-08 15:04:45.096 2767-2767/us.michaelchen.compasslogger W/System: ClassLoader referenced unknown path: /data/app/us.michaelchen.compasslogger-1/lib/arm
06-08 15:04:45.554 2767-2767/us.michaelchen.compasslogger I/ACRA: ACRA is enabled for us.michaelchen.compasslogger, initializing...
06-08 15:04:45.586 2767-2767/us.michaelchen.compasslogger D/AndroidRuntime: Shutting down VM
06-08 15:04:45.587 2767-2767/us.michaelchen.compasslogger E/ACRA: ACRA caught a RuntimeException for us.michaelchen.compasslogger
                                                                  java.lang.RuntimeException: Unable to start receiver us.michaelchen.compasslogger.receiver.BootReceiver: android.content.ReceiverCallNotAllowedException: BroadcastReceiver components are not allowed to register to receive intents
                                                                      at android.app.ActivityThread.handleReceiver(ActivityThread.java:2732)
                                                                      at android.app.ActivityThread.-wrap14(ActivityThread.java)
                                                                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1421)
                                                                      at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                      at android.os.Looper.loop(Looper.java:148)
                                                                      at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                      at java.lang.reflect.Method.invoke(Native Method)
                                                                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                   Caused by: android.content.ReceiverCallNotAllowedException: BroadcastReceiver components are not allowed to register to receive intents
                                                                      at android.app.ReceiverRestrictedContext.registerReceiver(ContextImpl.java:96)
                                                                      at android.app.ReceiverRestrictedContext.registerReceiver(ContextImpl.java:85)
                                                                      at us.michaelchen.compasslogger.utils.MasterSwitch.startAsynchronous(MasterSwitch.java:149)
                                                                      at us.michaelchen.compasslogger.utils.MasterSwitch.on(MasterSwitch.java:67)
                                                                      at us.michaelchen.compasslogger.receiver.BootReceiver.onReceive(BootReceiver.java:16)
                                                                      at android.app.ActivityThread.handleReceiver(ActivityThread.java:2725)
                                                                      at android.app.ActivityThread.-wrap14(ActivityThread.java) 
                                                                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1421) 
                                                                      at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                      at android.os.Looper.loop(Looper.java:148) 
                                                                      at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                      at java.lang.reflect.Method.invoke(Native Method) 
                                                                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
06-08 15:04:45.670 2767-2767/us.michaelchen.compasslogger I/Process: Sending signal. PID: 2767 SIG: 9
06-08 15:04:45.710 2823-2823/us.michaelchen.compasslogger:acra W/System: ClassLoader referenced unknown path: /data/app/us.michaelchen.compasslogger-1/lib/arm
06-08 15:04:45.872 2823-2823/us.michaelchen.compasslogger:acra W/System: ClassLoader referenced unknown path: /data/app/us.michaelchen.compasslogger-1/lib/arm
06-08 15:04:46.416 2823-2902/us.michaelchen.compasslogger:acra I/ACRA: Sending report /data/user/0/us.michaelchen.compasslogger/app_ACRA-approved/2016-06-08T15:04:45.602-07:00.stacktrace
06-08 15:04:46.795 2947-2947/us.michaelchen.compasslogger W/System: ClassLoader referenced unknown path: /data/app/us.michaelchen.compasslogger-1/lib/arm
06-08 15:04:47.053 2947-2947/us.michaelchen.compasslogger W/System: ClassLoader referenced unknown path: /data/app/us.michaelchen.compasslogger-1/lib/arm
06-08 15:04:47.077 2947-2947/us.michaelchen.compasslogger E/InstantRun: IO Error creating local socket at us.michaelchen.compasslogger
                                                                        java.io.IOException: Address already in use
                                                                            at android.net.LocalSocketImpl.bindLocal(Native Method)
                                                                            at android.net.LocalSocketImpl.bind(LocalSocketImpl.java:306)
                                                                            at android.net.LocalServerSocket.<init>(LocalServerSocket.java:48)
                                                                            at com.android.tools.fd.runtime.Server.<init>(Server.java:94)
                                                                            at com.android.tools.fd.runtime.Server.create(Server.java:88)
                                                                            at com.android.tools.fd.runtime.BootstrapApplication.onCreate(BootstrapApplication.java:358)
                                                                            at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1013)
                                                                            at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4707)
                                                                            at android.app.ActivityThread.-wrap1(ActivityThread.java)
                                                                            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1405)
                                                                            at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                            at android.os.Looper.loop(Looper.java:148)
                                                                            at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                            at java.lang.reflect.Method.invoke(Native Method)
                                                                            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
06-08 15:04:47.430 2947-2947/us.michaelchen.compasslogger I/ACRA: ACRA is enabled for us.michaelchen.compasslogger, initializing...
06-08 15:13:49.126 2947-2953/us.michaelchen.compasslogger W/art: Suspending all threads took: 20.254ms
06-08 15:14:38.740 2947-2953/us.michaelchen.compasslogger W/art: Suspending all threads took: 14.401ms```
``````

**Comment by safi-manar** [comment]
Even after commit 71503728253accac394dfc644b5f6f50ce700b18 , `MasterSwitch isRunning refactor cleanup.` the error persists after restarting.

**Build:** 71503728253accac394dfc644b5f6f50ce700b18
**UUID:** `78b0b07e-625e-47f4-b05c-af81bef632f4`

**Description:** The service does not restart upon a system restart and so data collection no longer persists.

**To Replicate:**

Install a fresh copy of 71503728253accac394dfc644b5f6f50ce700b18 and connect device to AndroidMonitor.
Proceed with the procedure of completing the survey.
Once alarms have been set, restart the device.
The error should occur when the device is unlocked. _The error does not appear in Android studio, as before, however. It is only visible in the crash logs._ The service is unable to restart and the device no longer collects and sends data.

**Error:**

```
"java.lang.RuntimeException: Unable to start receiver us.michaelchen.compasslogger.receiver.BootReceiver: android.content.ReceiverCallNotAllowedException: BroadcastReceiver components are not allowed to register to receive intents
    at android.app.ActivityThread.handleReceiver(ActivityThread.java:2732)
    at android.app.ActivityThread.-wrap14(ActivityThread.java)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1421)
    at android.os.Handler.dispatchMessage(Handler.java:102)
    at android.os.Looper.loop(Looper.java:148)
    at android.app.ActivityThread.main(ActivityThread.java:5417)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
Caused by: android.content.ReceiverCallNotAllowedException: BroadcastReceiver components are not allowed to register to receive intents
    at android.app.ReceiverRestrictedContext.registerReceiver(ContextImpl.java:96)
    at android.app.ReceiverRestrictedContext.registerReceiver(ContextImpl.java:85)
    at us.michaelchen.compasslogger.utils.MasterSwitch.startAsynchronous(MasterSwitch.java:228)
    at us.michaelchen.compasslogger.utils.MasterSwitch.on(MasterSwitch.java:80)
    at us.michaelchen.compasslogger.receiver.BootReceiver.onReceive(BootReceiver.java:16)
    at android.app.ActivityThread.handleReceiver(ActivityThread.java:2725)
    ... 8 more
android.content.ReceiverCallNotAllowedException: BroadcastReceiver components are not allowed to register to receive intents
    at android.app.ReceiverRestrictedContext.registerReceiver(ContextImpl.java:96)
    at android.app.ReceiverRestrictedContext.registerReceiver(ContextImpl.java:85)
    at us.michaelchen.compasslogger.utils.MasterSwitch.startAsynchronous(MasterSwitch.java:228)
    at us.michaelchen.compasslogger.utils.MasterSwitch.on(MasterSwitch.java:80)
    at us.michaelchen.compasslogger.receiver.BootReceiver.onReceive(BootReceiver.java:16)
    at android.app.ActivityThread.handleReceiver(ActivityThread.java:2725)
    at android.app.ActivityThread.-wrap14(ActivityThread.java)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1421)
    at android.os.Handler.dispatchMessage(Handler.java:102)
    at android.os.Looper.loop(Looper.java:148)
    at android.app.ActivityThread.main(ActivityThread.java:5417)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
"
```

**Comment by safi-manar** [comment]
Seems to be fixed by 75f1bf317a5a022e22498812728b8bd7c695d9ab .

http://stackoverflow.com/questions/24934260/intentreceiver-components-are-not-allowed-to-register-to-receive-intents-when

_answered Jul 24 '14 at 12:54 CommonsWare_

> Instead of:
> 
> c.registerReceiver(null, iFilter)
> use:
> 
> c.getApplicationContext().registerReceiver(null, iFilter)
> The Context that is passed to onReceive() is blocked from calling registerReceiver(), even with a null BroadcastReceiver.
